checkup: Add VirtualMachineInstance creation

orelmisan opened the pull request:
Create the VirtualMachineInstance object to be used as a test subject.

A follow-up PR will also wait for the VMI to boot.

The VMI object is created with a random suffix, in order to enable execution of multiple checkup instances simultaneously.

The e2e test was manually executed against an OpenShift 4.11 cluster, with a RT configured node.
The VMI had booted successfully.

orelmisan committed:
Introduce the vmi package

This package exposes the `New()` function and its option functions, which helps build the
VirtualMachineInstance object that will be used by the checkup.

Signed-off-by: Orel Misan <omisan@redhat.com>

orelmisan committed:
e2e: Add permissions to create a VMI

In order to allow creation of a VirtualMachineInstance object during
the checkup's setup phase - add the relevant Role and RoleBinding
objects.

Signed-off-by: Orel Misan <omisan@redhat.com>

orelmisan committed:
e2e, ConfigMap: Adjust input parameters

Use more realistic input values.

Signed-off-by: Orel Misan <omisan@redhat.com>

orelmisan committed:
client: Add CreateVirtualMachineInstance()

Add the CreateVirtualMachineInstance() function,
in order to create a new VirtualMachineInstance object.

This function uses a workaround to apply a context.
This workaround will be removed when kubevirt client-go
package will be upgraded to v0.59.0.

[1] https://github.com/kiagnose/kiagnose/pull/230
[2] https://github.com/kubevirt/kubevirt/issues/8885
[3] https://github.com/kubevirt/kubevirt/pull/8908

Signed-off-by: Orel Misan <omisan@redhat.com>

orelmisan committed:
main: Read namespace from file

Read the namespace on which the checkup runs.
Pass this namespace to mainflow's Run() function
in order to pass it to the Checkup object in a follow-up
commit.

Signed-off-by: Orel Misan <omisan@redhat.com>

orelmisan committed:
checkup, setup: Create a VirtualMachineInstance object

Create the VirtualMachineInstance object to be used
as a test subject.

A follow-up commit will also wait for the VMI to boot.

The VMI object is created with a random suffix, in order to
enable execution of multiple checkup instances simultaneously.

Signed-off-by: Orel Misan <omisan@redhat.com>

orelmisan committed:
mainflow: Create a context with timeout

Create a context with timeout, in order to respect
the user's timeout input.

Signed-off-by: Orel Misan <omisan@redhat.com>

EdDev commented on tests/checkup_test.go:
```diff
@@ -38,6 +38,8 @@ const (
 	testServiceAccountName                     = "rt-checkup-sa"
 	testKiagnoseConfigMapAccessRoleName        = "kiagnose-configmap-access"
 	testKiagnoseConfigMapAccessRoleBindingName = testKiagnoseConfigMapAccessRoleName
+	testKubeVirtRTCheckerRoleName              = "kubevirt-rt-checker"
```
nit: Why `checker` and not `checkup`? Same for all the rest.

orelmisan replied:
I've followed the Role name that we've used in the VM latency checkup:
https://github.com/kiagnose/kiagnose/tree/main/checkups/kubevirt-vm-latency#permissions

EdDev commented on tests/checkup_test.go:
```diff
@@ -134,6 +138,32 @@ func setupCheckupPermissions() {
+	Expect(err).NotTo(HaveOccurred())
+
+	DeferCleanup(func() {
+		err = client.RbacV1().Roles(testNamespace).Delete(context.Background(), testKubeVirtRTCheckerRoleName, metav1.DeleteOptions{})
```
It would be cleaner to take the name from `kubeVirtRTCheckerRole` object itself.

orelmisan replied:
Done.

orelmisan replied:
I'll create another PR to apply this change for the other objects.

EdDev commented on tests/checkup_test.go:
```diff
@@ -134,6 +138,32 @@ func setupCheckupPermissions() {
+	DeferCleanup(func() {
+		err = client.RbacV1().RoleBindings(testNamespace).Delete(
+			context.Background(),
+			testKubeVirtRTCheckerRoleBindingName,
```
It would be cleaner to take the name from `kubeVirtRTCheckerRoleBinding` object itself.

orelmisan replied:
Done.

EdDev commented on tests/checkup_test.go:
```diff
@@ -178,6 +208,40 @@ func newKiagnoseConfigMapAccessRoleBinding() *rbacv1.RoleBinding {
+	}
+}
+
+func newKubeVirtRTCheckerRoleBinding() *rbacv1.RoleBinding {
```
It makes sense (to me), to express the relation between the role and the role-binding, such that the role-binding will accept as input the role itself (or its name) with the SA.

In order to understand this relation, we need to travel through the constants, which makes this thing fragile IMO.

orelmisan replied:
Done.

orelmisan replied:
I'll create another PR to apply this change for the other objects.

EdDev commented on pkg/internal/checkup/vmi/vmi.go:
```diff
@@ -0,0 +1,289 @@
+)
+
+// Based on annotation names from:
+// https://github.com/cri-o/cri-o/blob/fa0fa5de1c17ddd7b6fcdbc030b6b571ce37e643/pkg/annotations/annotations.go
```
So this solution is CRI-O specific?

orelmisan replied:
AFAIK, the VMI manifest used here was tested on OpenShift Virtualization, which uses CRI-O.

I don't know if other container engines have similar knobs.

ormergi replied:
Is there a reference for using these annotations on Kubevirt docs?

Is there an API on Kubevirt to configure the VM similar to placing these annotations?

orelmisan replied:
There is this blog post about realtime VMs in KubeVirt:
https://kubevirt.io/2021/Running-Realtime-Workloads.html

It doesn't contain the cri-o annotations.

I also haven't found documentaion about these annotations in the [user guide](https://kubevirt.io/user-guide).

EdDev replied:
If the solution is so specific, I would make sure to describe it bodly on the README or any other document.

EdDev commented on pkg/internal/checkup/vmi/vmi.go:
```diff
@@ -0,0 +1,289 @@
+
+const Disable = "disable"
+
+type Builder struct {
```
I do not understand the need of this builder object.
The builder pattern was used in other places as a simple factory (without objects), why this is different here? Or more accurately, what does this object comes to solve exactly (compared to the flat function option)?

EdDev commented on pkg/internal/checkup/checkup.go:
```diff
@@ -47,3 +75,41 @@ func (c *Checkup) Teardown(ctx context.Context) error {
+	user: user`
+	)
+
+	newVMI := vmi.NewBuilder(randomizeName(VMINamePrefix)).
```
This is another builder pattern.
I do not understand why you had to re-invent another one when we already have one in use in the reference checkup.

I would expect to re-use that one instead. And if there is a good reason, to invent a new one.

I personally do not like this "train" pattern, it reminds Java builders.
While the K8S clientset uses this pattern, it uses it in a pretty fixed format (i.e. the order makes sense and is not endless).

orelmisan replied:
Changed the code to use a function, similar to the VM latency checkup.
Removed the builder object.

orelmisan commented:
[Change](https://github.com/kiagnose/kubevirt-rt-checkup/compare/94644848c68a04b994ba9fab6c35c8779d685173..a3e2fe7dba7c517ed25f81b6df88574d2cc09bc1): Answered review [comments](https://github.com/kiagnose/kubevirt-rt-checkup/pull/17#pullrequestreview-1249622774).

orelmisan commented:
[Change](https://github.com/kiagnose/kubevirt-rt-checkup/compare/a3e2fe7dba7c517ed25f81b6df88574d2cc09bc1..c8b874c0085ca82fb005d32f9676a39aafb631a3): Take the KubeVirtChecker and ServiceAccount names from the objects and not from the constants.

orelmisan commented:
[Change](https://github.com/kiagnose/kubevirt-rt-checkup/compare/c8b874c0085ca82fb005d32f9676a39aafb631a3..f18c035f2ed04cf1b20342c4a7db03b90306de59): Rebase

orelmisan commented:
[Change](https://github.com/kiagnose/kubevirt-rt-checkup/compare/f18c035f2ed04cf1b20342c4a7db03b90306de59..6e2d16ee6ab406a5cf7451029b0d37e5a62d83f1): Fixed a small nit.

RamLavi reviewed (dismissed):
/lgtm

EdDev commented on pkg/internal/checkup/vmi/vmi.go:
```diff
@@ -0,0 +1,237 @@
+
+func WithOwnerReference(ownerName, ownerUID string) Option {
+	return func(vmi *kvcorev1.VirtualMachineInstance) {
+		if ownerUID != "" && ownerName != "" {
```
Why do you need this condition? The caller is not controlling with what is called?

orelmisan replied:
this function will be called like this:
```go
vmi.New(testVMIName, 
...
vmi.WithOwnerReference(checkupConfig.PodName, checkupConfig.PodUID),
...
```
`checkupConfig.PodUID` value comes from downward API, and could be an empty string.

EdDev commented on pkg/internal/checkup/vmi/vmi.go:
```diff
@@ -0,0 +1,237 @@
+
+func WithNodeSelector(nodeName string) Option {
+	return func(vmi *kvcorev1.VirtualMachineInstance) {
+		if nodeName == "" {
+			return
+		}
```
Looks like defensive programing. I do not think we need it.

orelmisan replied:
This function is called like this:
```go
vmi.New(testVMIName, 
...
WithNodeSelector(checkupConfig.TargetNode),
...
```

`targetNode` is an optional field, and its default value is the empty string.
IMO we should not add a nodeSelector field to the VMI spec if we don't need it.

EdDev commented on pkg/internal/checkup/vmi/vmi.go:
```diff
@@ -0,0 +1,237 @@
+		},
+		Spec: kvcorev1.VirtualMachineInstanceSpec{
+			Domain: kvcorev1.DomainSpec{},
+		},
```
Is this needed? If it is not a pointer, it is implicitly initialized as such, no?

orelmisan replied:
You are correct.
I've removed the explicit initialization of `Spec`.

EdDev commented on pkg/internal/checkup/vmi/vmi_test.go:
```diff
@@ -0,0 +1,305 @@
+ *
+ */
+
+package vmi_test
```
Testing the vmi spec feels like an overkill.
I would focus on testing something at a higher level than this, and if there is none, just keep it covered by the e2e tests.

But it up to you, in the end you will need to maintain this test code.

orelmisan replied:
I'll keep them for the time being, and consider their removal when the checkup tests will cover some of the VMI spec creation (such as nodeSelector etc).

EdDev reviewed (commented):
The comments I left can be consider non-blocking.
If you prefer to continue as is, it is fine.

orelmisan commented:
[Change](https://github.com/kiagnose/kubevirt-rt-checkup/compare/6e2d16ee6ab406a5cf7451029b0d37e5a62d83f1..bad0b2140ac45d42c2a2304bc2c550e8d80786ea): Removed explicit initialization of VMI.Spec.
Removed redundant white space before cloudInit data.
